pyro/distributions/ordered_logistic.py

```diff
@@ -0,0 +1,57 @@
+import torch
+from torch.distributions import constraints
+from pyro.distributions.torch import Categorical
+
+
+class OrderedLogistic(Categorical):
+    """
+    Alternative parametrization of the distribution over a categorical variable.
+
+    Instead of the typical parametrization of a categorical variable in terms
+    of the probability mass of the individual categories ``p``, this provides an
+    alternative that is useful in specifying ordered categorical models. This
+    accepts a vector of ``cutpoints`` which are an ordered vector of real numbers
+    denoting baseline cumulative log-odds of the individual categories, and a
+    model vector ``predictors`` which modifies the baselines for each sample
+    individually.
+
+    These cumulative log-odds are then transformed into a discrete cumulative
+    probability distribution, that is finally differenced to return the probability
+    mass matrix ``p`` that specifies the categorical distribution.
+
+    :param Tensor predictors: A tensor of predictor variables of arbitrary
+        shape. The output shape of non-batched samples from this distribution will
+        be the same shape as ``predictors``.
+    :param Tensor cutpoints: A tensor of cutpoints that are used to determine the
+        cumulative probability of each entry in ``predictors`` belonging to a
+        given category. Must be 1-dimensional and monotonically increasing.
+    """
+
+    support = constraints.nonnegative_integer
+    arg_constraints = {
+        "predictors": constraints.real,
+        "cutpoints": constraints.real_vector,
+    }
+    has_rsample = False
+
+    def __init__(self, predictors, cutpoints):
+        assert cutpoints.ndim == 1, "cutpoints must be 1d vector"
+        assert predictors.ndim >= 1, "predictors must have at least 1 dimension"
```

Comment:
Thanks for supporting `batch_shape` for `predictors`! Could you also support batch_shape for `cutpoints`? I would expect those parameters have shape
```
cutpoints.shape == batch_shape + event_shape
predictors.shape == batch_shape
```
where `event_shape = cutpoints.shape[-1:]` and `batch_shape` can be empty. You can broadcast those parameters as in [MVN implementation](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L130).

Reply:
I think I found something a little more flexible and intuitive than fixing `cutpoints` and `predictors` to have the same `batch_shape` - as long as the first `N = cutpoints.ndim - 1` dimensions are broadcastable to the last `N` dimensions of `predictors`, then we can just use normal broadcasting semantics to handle it by appending an event dim to `predictors`:

```python
q = torch.sigmoid(cutpoints - predictors.view(predictors.shape + (1,)))
```

Check out the test I wrote for it: https://github.com/pyro-ppl/pyro/pull/2651/files#diff-f0cc7966b175f63150546ace313d0179R31-R42 - does that make sense?

Reply:
It makes sense to me. Thanks!